### DISCUSSION TITLE
Community Platformio support for Arduino core 3.x.x

### DISCUSSION THREAD

**Discussion by Jason2866** [espressif/arduino-esp32#10039] [General]
Since it is now clear that there will be no official support for Arduino core 3.0.x from Platformio team, (Issue No. 1225 in repo https://github.com/platformio/platform-espressif32/issues) i decided to fork the needed repo(s) and build a community version to support core 3.x (started with core 3.0.3)

**Arduino core 3.1.1 is available** https://github.com/pioarduino/platform-espressif32

```
[env:stable]
platform = https://github.com/pioarduino/platform-espressif32/releases/download/stable/platform-espressif32.zip
board = ...
...
```

Since the team of this community version is small ~~currently just me ;-)~~, there is at the moment ~~very~~ **limited support** possible! Of course the issue section will be open for providing problems!

Since the team is currently small ~~a "one man show"~~, it would be cool to find helping people to make this to a real community work.

~~Every user of Platformio using Arduino espressif32 can do the following and being a great help!~~

~~**- Priority No.1 help:**  Done! Thx for the help. The left missing I have downloaded myself.~~

Further help for:
- Writing a small? Documentation how to use the fork version
- Documentation of the build process (Arduino Lib Builder) to compile Arduino libs useable with Platformio
- Guide how to build a IDF release package (from source) to be used in Platformio (actually IDF 5.3.2)
- Coding: Implement a patch routine to apply (during install) the patches from Arduino Lib Builder to IDF
- Setup and maintain Docus on Community repo (preferred choice MkDocs) -> Done
- Describe how to compile Arduino as an component of IDF 5.3.x
- Providing support -> make it possible to open Github Discussions and a Discord server -> Done
~~- Programming in Python, no fear no big jobs ;-)~~
- Coding /administrative work: Use espressif / community repo as source for tools and toolchains (remove dependency to Platformio registry) -> DONE
- Adding boards support
- Review PRs, especially Boards support
~~- Midterm goal: provide additionally a stripped down low resource framework (Rainmaker, TfLite, Insights and other fat and seldom used components removed)~~ Not done, replaced by Hybrid Compile
~~- Long term, stretch goal (maybe impossible?): Modular built version (specify wanted IDF/Arduino components and configs)~~ Documentation for compile mode "Hybrid compile"
- and other things, currently not coming in mind...

Would be great to get feedback here and where you think you could help

Great response!

Discord Server for coordinating and early support:
The Discord invite link is limited by default :-(
This one should not expire
[https://discord.gg/Nutz9crnZr](https://discord.gg/Nutz9crnZr)

identified tasks to work on:
~~- how to install tools from github~~ @vortigont Thx. Done.

~~ready to start to work on:~~
~~- Docu for using pioarduino -> repo ready to be filled https://github.com/pioarduino/docs Who?~~ Done

**Question: Is it fine for all, if I hide posts which are outdated or solved? Goal -> Better visibility
Please post if you against this way. Thx.**

EDIT: This first post is edited, outdated todo and infos are removed. Thx to all!

**Comment by vortigont** [comment]
Nice! Would be more than happy to help with anything!

**Comment by sivar2311** [comment]
> Nice! Would be more than happy to help with anything!

Same here.
@Jason2866 Let me know what I can do

**Comment by cziter15** [comment]
I think I can help with pretty much anything you've listed ;)

**Comment by Jason2866** [comment]
Here we go with Arduino Core 3.0.3 Give it a try with this platform

```
[env:stable]
platform = https://github.com/pioarduino/platform-espressif32/releases/download/51.03.03/platform-espressif32.zip
board = ...
...
```
The number `51.03.03` in the URL has the logic `51` = IDF 5.1 and `03.03` stands for core 3.0.3
The source is here: https://github.com/pioarduino/platform-espressif32

**Comment by Jason2866** [comment]
@fredlcore @fundix @ricardoquesada @hitecSmartHome

Arduino as an component works. Use the usual known setup
```
[env:stable]
framework = arduino, espidf
platform = https://github.com/pioarduino/platform-espressif32/releases/download/51.03.03/platform-espressif32.zip
board = ...
```

**Reply by fredlcore** [reply]
Awesome, thanks!

**Reply by hitecSmartHome** [reply]
Wow :o will check it out! Thank you

**Comment by ForrestFire0** [comment]
Happy to help in any way I can. @Jason2866   Although my reasons are selfish...

**Reply by Jason2866** [reply]
Thats the best motivation :-)

**Comment by vortigont** [comment]
do we have a discord server to join? Or maybe a Telegram channel?

**Comment by Jason2866** [comment]
@vortigont ~~Not yet, for now we can use the Tasmota Discord server.~~
yes we should open soon a Discord server for. Done, Discord Server is pioarduino
[https://discord.gg/Nutz9crnZr](https://discord.gg/Nutz9crnZr)

**Reply by szerwi** [reply]
@Jason2866 Invitation to the Discord server is not working. Could you send me new link?

**Reply by Jason2866** [reply]
The default setting is expire. Who ever made this silly default...
Hopefully this one lasts [https://discord.gg/Nutz9crnZr](https://discord.gg/Nutz9crnZr)

**Comment by mathieucarbou** [comment]
If needed, to fasten things:

put in pio file:
```
[env:pioarduino-esp32dev]
platform = https://github.com/pioarduino/platform-espressif32/releases/download/51.03.03/platform-espressif32.zip
board = esp32dev

[env:pioarduino-esp32-s2]
platform = https://github.com/pioarduino/platform-espressif32/releases/download/51.03.03/platform-espressif32.zip
board = esp32-s2-saola-1

[env:pioarduino-esp32-s3]
platform = https://github.com/pioarduino/platform-espressif32/releases/download/51.03.03/platform-espressif32.zip
board = esp32-s3-devkitc-1

[env:pioarduino-esp32-c3]
platform = https://github.com/pioarduino/platform-espressif32/releases/download/51.03.03/platform-espressif32.zip
board = esp32-c3-devkitc-02

[env:pioarduino-esp32-c6]
platform = https://github.com/pioarduino/platform-espressif32/releases/download/51.03.03/platform-espressif32.zip
board = esp32-c6-devkitc-1
```
run:
```bash
# cleanup existing packages
rm ~/.platformio/packages
# build
for e in `pio project config --json-output | jq -cr '.[][0] | select(startswith("env:pioarduino-")) | .[4:]'`; do pio run -e $e; done;
# compress
cd ~/.platformio/packages
ls | grep -v framework | grep -v zip | xargs -n1 -I {} zip -s 100m -r {}.zip {}
# double check your architecture:
uname -a
```
Then move all the zip and zxx files in your fork of https://github.com/pioarduino/registry, in the right folder.
Push, PR, :wink:

This accounts for about 750Mb in my case for macos_arm64.

**Comment by Jason2866** [comment]
@brentru Saw https://github.com/espressif/arduino-esp32/discussions/10061#discussioncomment-10126930 idk if I manage to join the virtual meeting. May I can answer your questions here? Or contact me on pioarduino Discord. espressif is not related to pioarduino and can't answer questions regarding too.

**Reply by brentru** [reply]
Okay! I will remove this topic from the discussion and contact you via Discord.

**Comment by Xylopyrographer** [comment]
@Jason2866 : Regarding the note above

> ready to start to work on:
> 
> Docu for using pioarduino -> repo ready to be filled https://github.com/pioarduino/docs Who?

What docs in addition to what's over on platform.io are needed?

I think I saw in here somewhere MkDocs is the tool of choice? If so using their [readthedocs]( https://www.mkdocs.org/user-guide/choosing-your-theme/#readthedocs) theme?

**Reply by Jason2866** [reply]
> @Jason2866 : Regarding the note above
> 
> > ready to start to work on:
> > Docu for using pioarduino -> repo ready to be filled https://github.com/pioarduino/docs Who?
> 
> What docs in addition to what's over on platform.io are needed?
> 
> I think I saw in here somewhere MkDocs is the tool of choice? If so using their [readthedocs](https://www.mkdocs.org/user-guide/choosing-your-theme/#readthedocs) theme?

@Xylopyrographer Currently the only difference i see is the different Platform used and it is a link to a zip file. I have not really knowledge regarding mkdocs. Overall imho to ask the developer what's needed is not optimal. He just know how it all works. Would it be possible you just start? If you have questions contact me at Discord.

**Thx a lot for helping!**

**Reply by Xylopyrographer** [reply]
Gotcha. I'll ponder for a bit & then raise a PR for the `/docs` repository.

FWIW, IMHO, MkDocs is a better tool for community supported repositories. 👍

**Comment by Hedda** [comment]
FYI, it looks like we can see practically the same situation between PlatformIO and Raspberry Pi Foundation or Raspberry Pi Ltd.

* https://github.com/platformio/platform-raspberrypi/pull/36

PlatformIO is now also refusing community contributions to improve its platform-raspberrypi library, so it has stalled now too:

  * https://github.com/platformio/platform-raspberrypi/pull/70

Because of this users are forced to move to forks or alternative Arduino-Pico cores in order to use the new RP2350 MCU/SoC. Ex:

* https://github.com/maxgerhardt/platform-raspberrypi/

* https://github.com/earlephilhower/arduino-pico/

PS: I understand PlatformIO is a company that need income but I do not see how this practice will benefit them in the long run.
